# STAT 207 Homework 10 [25 points]

## Logistic Regression, Classifiers, & Project 3

Due: Friday, April 19, end of day (11:59 pm CT)

<hr>

## Imports 

Run the following code cell to import the necessary packages into the file.  You may import additional packages, as needed for this assignment.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns; sns.set()
import statsmodels.formula.api as smf
from sklearn.model_selection import train_test_split

## Case Study: Heart Disease

Can we identify patients who have heart disease based on other patient characteristics?

We will use the following information about patients:

- **age**: age
- **sex**: sex, 1 = male, 0 = female
- **cp**: chest pain type, 1 = typical angina, 2 = atypical angina, 3 = non-anginal pain, 4 = asymptomatic
- **trestbps**: resting blood pressure (in mm Hg)
- **chol**: serum cholesterol in mg/dl
- **fbs**: whether the fasting blood sugar is above 120 mg/dl (1 = true, 0 = false)
- **restecg**: resting electrocardiographic results (0 = normal, 1 = ST-T wave abnormality, 2 = left ventricular hypertrophy)
- **thalach**: maximum heart rate achieved
- **exang**: exercise induced angina (1 = yes, 0 = no)
- **oldpeak**: ST depression induced by exercise relative to rest
- **slope**: the slope of the peak exercise ST segment (1 = upsloping, 2 = flat, 3 = downsloping)
- **ca**: number of major vessels (0-3) colored by flouroscopy (for calcification of vessels)
- **thal**: number of nuclear stress test (3 = normal, 6 = fixed defect, 7 = reversible defect)
- **num**: target number of progression of heart disease (1+ indicates at least 50% narrowing
- **y**: our target, generated below, for the presence of heart disease (1 = present, 0 = absent)

The data originally came from https://archive.ics.uci.edu/dataset/45/heart+disease.

Run the code in the cell below to read in the data for this document.  The data is saved as `df` with this code.  

In [2]:
df = pd.read_csv('processed_cleveland.csv', na_values = ['?'])
df = df.dropna()
df['y'] = 1 * (df['num'] > 0)
df['sex'] = df['sex'].replace({1: 'male', 0: 'female'})
df['cp'] = df['cp'].replace({1: 'typical', 2:'atypical', 3: 'non-anginal', 4: 'asymptomatic'})
df['fbs'] = df['fbs'].replace({1: 'high', 0: 'typical'})
df['restecg'] = df['restecg'].replace({0: 'normal', 1:'ST-T wave abnormality', 2: 'left ventricular hypertrophy'})
df['exang'] = df['exang'].replace({0: 'no', 1: 'yes'})
df['slope'] = df['slope'].replace({1:'upsloping', 2:'flat', 3:'downsloping'})
df['thal'] = df['thal'].replace({3: 'normal', 6: 'fixed defect', 7:'reversible defect'})

In [3]:
df.head()

age     sex            cp  trestbps  chol      fbs  \
0   63    male       typical       145   233     high   
1   67    male  asymptomatic       160   286  typical   
2   67    male  asymptomatic       120   229  typical   
3   37    male   non-anginal       130   250  typical   
4   41  female      atypical       130   204  typical   

                        restecg  thalach exang  oldpeak        slope   ca  \
0  left ventricular hypertrophy      150    no      2.3  downsloping  0.0   
1  left ventricular hypertrophy      108   yes      1.5         flat  3.0   
2  left ventricular hypertrophy      129   yes      2.6         flat  2.0   
3                        normal      187    no      3.5  downsloping  0.0   
4  left ventricular hypertrophy      172    no      1.4    upsloping  0.0   

                thal  num  y  
0       fixed defect    0  0  
1             normal    2  1  
2  reversible defect    1  1  
3             normal    0  0  
4             normal    0  0

## 0. Prepare the Data [0.5 points]

Split the data into a training and a test data, using 20% as our test data and a random state of 1103.  We will use our training data exclusively for the first three questions of the assignment.  We won't turn to the test data until the third question.

In [4]:
df_train, df_test = train_test_split(df, test_size = 0.2, random_state = 1103)

## 1. Understand a Logistic Regression Model [2 points]

We know that we would like to predict the presence of heart disease (y).  First, we'll consider a simple model that uses age to predict the presence of heart disease.

**a)** Below, fit the model predicting the presence of heart disease from age.  Print the summary of the model.

**b)** Generate the fitted vs. deviance residual plot for this model.

## 2. Fitting a More Complex Model [1 point]

We'd like to explore how effective advanced tests are at recognizing the risk of heart disease.  We'll fit a second model predicting y from age, sex, restecg, thalach, exang, oldpeak, slope, ca, and thal.  Be sure to print the output of your model below.

## 3. Build a Classifier [1.5 points]

Using the model from Question 2, build a classifier to our training data using a predictive probability threshold of 0.75.  Then, generate a confusion matrix from this model.

## 4. Test Data [1 point]

Finally, apply your logistic regression model from **Question 2** to your test data.  Calculate the predicted probabilities on your test data.

## 5. Starting the Project [4 points]

The last four points on this assignment will be earned by starting on the project.  These points serve as an incentive for starting the project early.  This question will be graded for completion rather than correctness, so no feedback will be given on your initial submissions for the project.  Note that if you are completing the project as a group, you may all submit the same code for this question.  However, each of your should submit that code on your own homework assignment and post it yourself to GitHub.

Be sure that you put a copy of your project data into this homework folder, so your code will run.

You may modify your project after you submit the portions of this assignment.

**a)** Read in your data.  Then, split your data into a testing and training set.

**b)** Fit a logistic model to your data, using at least three predictor variables.  Print the summary of this model.

**c)** Generate an ROC curve for the logistic regression model fit in **5b**.

Remember to keep all your cells and hit the save icon above periodically to checkpoint (save) your results on your local computer. Once you are satisified with your results restart the kernel and run all (Kernel -> Restart & Run All). **Make sure nothing has changed**. Checkpoint and exit (File -> Save and Checkpoint + File -> Close and Halt). Follow the instructions on the Homework 10 Canvas Assignment to submit your notebook to GitHub.